module_utils/ansible_nailgun_cement.py

```diff
@@ -266,6 +266,16 @@ def naildown_entity(entity_class, entity_dict, entity, state, module, check_miss
     return changed, changed_entity
 
 
+def search_entities_json(entity_class, search):
+    """ Find entities using search """
+    return entity_class().search_json(
+        query={'search': search,
+               # This is a hack to work around pagination in API-V2
+               # See Redmine: #21800
+               'per_page': 2 << 31}
+    )
```

Comment:
Due to the structure of the current foreman API, this will not always return all fields of the entities. A subsequent `read` is needed in some cases.

Reply:
Initially, I just want to return data from the Foreman API. If we want to return nested objects we can do that in another PR, I'm especially reluctant as it could be very slow, especially for a large number of hosts.

Reply:
This is not about nested entities, but about fields that are in the detailed json view (`show.json.rabl`), but not the json overview (`base.json.rabl`).
Take for example products.

Reply:
sounds good there...

Reply:
I choose not to do this for now, I think we're better off always returning the same datatype (a list). I have renamed the module to `foreman_search_facts` to better reflect what it is doing. I think creating a `foreman_read_facts` modules or something alike would be a better way to solve that use case. Are you okay with that?

Reply:
Would the read facts also do the search? Then we could just add a boolean `read_full` here. Would be hard to document properly either way.
Anyway, we can do that later.

Reply:
> Would the read facts also do the search?

No - I imagine it would take an ID or name (something unique about that entity). If required users can use this module to find the ID and `foreman_read_facts` to get more info.

Reply:
Sounds good.